include/unifex/async_generator.hpp

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// A lot of stuff from cppcoro
+
+#pragma once
+
+#if !UNIFEX_NO_COROUTINES
+
+#  include <unifex/any_scheduler.hpp>
+#  include <unifex/async_scope.hpp>
+#  include <unifex/await_transform.hpp>
+#  include <unifex/create.hpp>
+#  include <unifex/defer.hpp>
+#  include <unifex/inline_scheduler.hpp>
+#  include <unifex/just_void_or_done.hpp>
+#  include <unifex/task.hpp>
+#  include <unifex/with_scheduler_affinity.hpp>
+
+#  include <optional>
+
+#  include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+template <typename T>
+class async_generator;
+
+namespace detail {
+
+// TODO: read_scheduler should be generalized to the read() sender factory from
+// p2300, so we can do things like read(get_scheduler), read(get_stop_token),
+// etc.
+struct _read_scheduler_sender {
+  template <
+      template <typename...>
+      class Variant,
+      template <typename...>
+      class Tuple>
+  using value_types = Variant<Tuple<any_scheduler>>;
+
+  template <template <typename...> class Variant>
+  using error_types = Variant<std::exception_ptr>;
+
+  static constexpr bool sends_done = false;
+
+  static constexpr auto blocking = blocking_kind::always_inline;
+
+  template <class R>
+  struct _op_state {
+    R r_;
+
+    friend void tag_invoke(tag_t<start>, _op_state& op) noexcept {
+      try {
+        set_value(std::move(op.r_), get_scheduler(op.r_));
+      } catch (...) {
+        set_error(std::move(op.r_), std::current_exception());
+      }
+    }
+  };
+
+  template <typename R>
+  friend _op_state<std::decay_t<R>>
+  tag_invoke(tag_t<connect>, _read_scheduler_sender, R&& r) {
+    return {std::forward<R>(r)};
+  }
+};
+
+struct _read_scheduler_t {
+  constexpr _read_scheduler_sender operator()() const noexcept { return {}; }
+};
+
+inline constexpr _read_scheduler_t read_scheduler{};
+
+template <typename Promise>
+struct reschedule_receiver {
+  std::optional<typename Promise::value_type> value_;
```

Comment:
See below for why I think this is safe.
```suggestion
  typename Promise::value_type* value_;
```